File: Python/235.Lowest Common Ancestor of a Binary Search Tree.py
#返回二叉搜素树两个节点的最小共同祖先
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        cur = root
        if p.val>q.val: q , p = p , q
        while cur:
            if cur==p or cur==q:
                return cur
            if cur.val>p.val and cur.val<q.val:
                return cur
            elif cur.val>q.val:
                cur = cur.left
            else:
                cur = cur.right

t1 = TreeNode(1)
t2 = TreeNode(2)
t3 = TreeNode(3)
t4 = TreeNode(4)
t5 = TreeNode(5)
t6 = TreeNode(6)
t7 = TreeNode(7)
t8 = TreeNode(8)
t9 = TreeNode(9)
t6.left = t4
t6.right = t8
t4.left = t2
t4.right = t5
t2.left = t1
t2.right = t3
t8.left =  t7
t8.right = t9

s = Solution()
print(s.lowestCommonAncestor(t6,t3,t2).val)

File: C++/Tree/235.Lowest Common Ancestor of a Binary Search Tree.cpp
/*
求二叉搜索树中，两个结点的最低公共祖先结点。
*/
/*
思路：找出p、q中大的那个，从根结点开始，如果结点值大于大的那个，那么向左走，
	  如果结点值小于小的那个，那么向右走。如果在二者中间，跳出循环，返回该结点。
*/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(p->val > q->val)		std::swap(p,q);   //保证p<q
        TreeNode *cur = root;
        while(1)
        {
        	if(cur->val>q->val)	cur = cur->left;
        	else if(cur->val<p->val)	cur = cur->right;
        	else break;
        }
        return cur;
    }
};
int main()
{
	Solution s;
	TreeNode n1(1),n2(2),n3(3),n4(4),n5(5),n6(6);
	n4.left = &n3, n4.right = &n5;
	n3.left = &n1, n1.right = &n2;
	n5.right = &n6;
	TreeNode* out = s.lowestCommonAncestor(&n4,&n1,&n3);
	cout<<out->val<<endl;
}

